src/box/lua/config/utils/schema.lua

```diff
@@ -1389,6 +1389,133 @@ end
+-- node using preprocess_leave() and step down again into another
+-- child node.
+--
+-- At each given point the top annotation frame represents
+-- annotations merged from the root node down to the given one.
+-- If there are same named annotations on the path, then one from
+-- a descendant node is preferred.
+local function preprocess_enter(ctx, schema)
+    assert(ctx.annotation_stack ~= nil)
+    local non_annotation_keys = {
+        type = true,
+        fields = true,
+        key = true,
+        value = true,
+        items = true,
+    }
+
+    -- There are known annotations that barely has any sense in
+    -- context of descendant schema nodes. Don't track them.
+    local ignored_annotations = {
+        allowed_values = true,
+        validate = true,
+        default = true,
+        apply_default_if = true,
+    }
+
+    local frame = table.copy(ctx.annotation_stack[#ctx.annotation_stack] or {})
+    for k, v in pairs(schema) do
+        if not non_annotation_keys[k] and not ignored_annotations[k] then
+            frame[k] = v
+        end
+    end
+    table.insert(ctx.annotation_stack, frame)
+end
+
+-- Step up from a schema node.
+--
+-- Returns the computed fields for the node that we're leaving.
+--
+-- See preprocess_enter() for details.
+local function preprocess_leave(ctx)
+    return {
+        annotations = table.remove(ctx.annotation_stack),
+    }
+end
+
+-- The function prepares the given schema node tree in the
+-- following way.
+--
+-- * The schema node is copied.
+-- * All its descendant nodes are prepared using this algorithm
+--   (recursively).
+-- * Computed fields are calculated and stored in the copied node.
+-- * The copied node is the result.
+--
+-- The sketchy structure of the returned schema node is the
+-- following.
+--
+-- {
+--     <..fields..>
+--     computed = {
+--         annotations = <...>,
+--     },
+-- }
+--
+-- At now there is only one computed field: `annotations`.
+--
+-- The `annotations` field contains all the annotations merged
+-- from the root schema node down to the given one. If the same
+-- annotation is present in an ancestor node and in an descendant
+-- node, the latter is preferred.
+--
+-- Design details
+-- --------------
+--
+-- The copying is performed to decouple schema node definitions
+-- provided by a caller from the schema nodes stored in the schema
+-- object. It allows to modify them without modifying caller's own
+-- objects.
+--
+-- It is especially important if parts of one schema are passed
+-- as schema node definitions to create another schema.
+--
+-- This case also motivates to store all the computed fields in
+-- the `computed` field. This way it is easier to guarantee that
+-- all the computed fields are stripped down from the original
+-- schema nodes and don't affect the new schema anyhow.
+local function preprocess_schema(schema, ctx)
+    local res = table.copy(schema)
+
+    res.computed = nil
```

Comment:
I just think that it is less error-prone than handling it in `preprocess_enter()`, which may later gain more preprocessing logic. Added some comments. I hope now it is more clear.

```diff
diff --git a/src/box/lua/config/utils/schema.lua b/src/box/lua/config/utils/schema.lua
index d01351858f..8f5a74fb59 100644
--- a/src/box/lua/config/utils/schema.lua
+++ b/src/box/lua/config/utils/schema.lua
@@ -1490,9 +1490,29 @@ end
 -- all the computed fields are stripped down from the original
 -- schema nodes and don't affect the new schema anyhow.
 local function preprocess_schema(schema, ctx)
+    -- A schema node from another (already constructed) schema may
+    -- be used in the schema that is currently under construction.
+    --
+    -- Since we're going to modify the schema node, we should copy
+    -- it beforehand. Our modifications must not affect another
+    -- schema.
     local res = table.copy(schema)
 
+    -- Eliminate previously computed values if any.
+    --
+    -- The past values were computed against node's place in
+    -- another schema, so it must not influence the node in its
+    -- current place in the given schema.
+    --
+    -- This field is rewritten at end of the function, but it
+    -- anyway should be stripped before going to
+    -- preprocess_enter(). Otherwise it would be taken as an
+    -- annotation.
+    --
+    -- (It can be just ignored in preprocess_enter(), but dropping
+    -- it beforehand looks less error-prone.)
     res.computed = nil
+
     preprocess_enter(ctx, res)
 
     -- luacheck: ignore 542 empty if branch
```